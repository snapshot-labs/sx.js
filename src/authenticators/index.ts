import vanillaAuthenticator from './vanilla';
import ethSigAuthenticator from './ethSig';
import * as utils from '../utils';
import type { Authenticator } from '../types';

const authenticators = {
  '0xb32364e042cb948be62a09355595a4b80dfff4eb11a485c1950ace70b0e83500': vanillaAuthenticator,
  '0x06aac1e90da5df37bd59ac52b638a22de15231cbb78353b121df987873d0f369': ethSigAuthenticator
};

export function getAuthenticator(address: string): Authenticator | null {
  const authenticator = authenticators[utils.encoding.hexPadRight(address)];
  if (!authenticator) return null;

  return authenticator;
}
