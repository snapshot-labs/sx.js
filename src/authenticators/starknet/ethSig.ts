import { Call, CallData, shortString } from 'starknet';
import EthSigAuthenticatorAbi from './abis/EthSigAuthenticator.json';
import { getChoiceEnum } from '../../utils/starknet-enums';
import {
  Authenticator,
  Envelope,
  Propose,
  Vote,
  UpdateProposal,
  ProposeCallArgs,
  VoteCallArgs,
  UpdateProposalCallArgs
} from '../../types';

const callData = new CallData(EthSigAuthenticatorAbi);

export default function createEthSigAuthenticator(): Authenticator {
  return {
    type: 'ethSig',
    createProposeCall(envelope: Envelope<Propose>, args: ProposeCallArgs): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_propose', [
        r,
        s,
        v,
        space,
        args.author,
        shortString.splitLongString(args.metadataUri),
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        args.strategiesParams,
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_propose',
        calldata: compiled
      };
    },
    createVoteCall(envelope: Envelope<Vote>, args: VoteCallArgs): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_vote', [
        r,
        s,
        v,
        space,
        args.voter,
        args.proposalId,
        getChoiceEnum(args.choice),
        args.votingStrategies.map(strategy => ({
          index: strategy.index,
          params: strategy.params
        })),
        shortString.splitLongString(args.metadataUri)
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_vote',
        calldata: compiled
      };
    },
    createUpdateProposalCall(
      envelope: Envelope<UpdateProposal>,
      args: UpdateProposalCallArgs
    ): Call {
      const { space, authenticator } = envelope.data;

      if (!envelope.signatureData?.signature) {
        throw new Error('signature is required for this authenticator');
      }

      if (!envelope.signatureData?.message) {
        throw new Error('message is required for this authenticator');
      }

      const [r, s, v] = envelope.signatureData.signature;

      const compiled = callData.compile('authenticate_update_proposal', [
        r,
        s,
        v,
        space,
        {
          address: args.author
        },
        args.proposalId,
        {
          address: args.executionStrategy.address,
          params: args.executionStrategy.params
        },
        shortString.splitLongString(args.metadataUri),
        envelope.signatureData.message.salt
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_update_proposal',
        calldata: compiled
      };
    }
  };
}
