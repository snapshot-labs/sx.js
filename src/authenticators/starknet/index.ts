import createVanillaAuthenticator from './vanilla';
import createEthSigAuthenticator from './ethSig';
import createEthTxAuthenticator from './ethTx';
import createStarkSigAuthenticator from './starkSig';
import createStarkTxAuthenticator from './starkTx';
import { hexPadLeft } from '../../utils/encoding';
import type { Authenticator, NetworkConfig } from '../../types';

export function getAuthenticator(
  address: string,
  networkConfig: NetworkConfig
): Authenticator | null {
  const authenticator = networkConfig.authenticators[hexPadLeft(address)];
  if (!authenticator) return null;

  if (authenticator.type === 'vanilla') {
    return createVanillaAuthenticator();
  }

  if (authenticator.type === 'ethSig') {
    return createEthSigAuthenticator();
  }

  if (authenticator.type === 'ethTx') {
    return createEthTxAuthenticator();
  }

  if (authenticator.type === 'starkSig') {
    return createStarkSigAuthenticator();
  }

  if (authenticator.type === 'starkTx') {
    return createStarkTxAuthenticator();
  }

  return null;
}
