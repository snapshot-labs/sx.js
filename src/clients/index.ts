export { StarkNetTx } from './starknet/starknet-tx';
export { EthereumTx } from './starknet/ethereum-tx';
export { EthereumSig } from './starknet/ethereum-sig';
export { StarkNetSig } from './starknet/starknet-sig';
export { L1Executor } from './starknet/l1-executor';
export { EthereumTx as EvmEthereumTx } from './evm/ethereum-tx';
export { EthereumSig as EvmEthereumSig } from './evm/ethereum-sig';
