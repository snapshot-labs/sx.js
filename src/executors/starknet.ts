import { createStarknetExecutionParams } from '../utils/encoding';
import type { Call } from 'starknet';

export default function createStarknetExecutor() {
  return {
    type: 'starknet',
    getExecutionData(executorAddress: string, calls: Call[]) {
      return {
        executor: executorAddress,
        executionParams: createStarknetExecutionParams(calls)
      };
    }
  };
}
