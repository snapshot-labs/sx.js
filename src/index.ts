export * as clients from './clients';
export * as utils from './utils';
export { getExecutionData } from './executors';
export * from './strategies';
export * from './networks';
export * from './types';
