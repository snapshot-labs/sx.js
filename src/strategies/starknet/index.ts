import createVanillaStrategy from './vanilla';
import createMerkleWhitelistStrategy from './merkleWhitelist';
import createErc20VotesStrategy from './erc20Votes';
import { hexPadLeft } from '../../utils/encoding';
import type { Strategy, NetworkConfig } from '../../types';

export function getStrategy(address: string, networkConfig: NetworkConfig): Strategy | null {
  const strategy = networkConfig.strategies[hexPadLeft(address)];
  if (!strategy) return null;

  if (strategy.type === 'vanilla') {
    return createVanillaStrategy();
  }

  if (strategy.type === 'whitelist') {
    return createMerkleWhitelistStrategy();
  }

  if (strategy.type === 'erc20Votes') {
    return createErc20VotesStrategy();
  }

  return null;
}
