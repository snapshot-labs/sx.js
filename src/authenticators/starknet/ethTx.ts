import { Call, CallData } from 'starknet';
import EthTxAuthenticatorAbi from './abis/EthTxAuthenticator.json';
import {
  Authenticator,
  Envelope,
  Propose,
  Vote,
  UpdateProposal,
  ProposeCallArgs,
  VoteCallArgs,
  UpdateProposalCallArgs
} from '../../types';

const callData = new CallData(EthTxAuthenticatorAbi);

export default function createEthTxAuthenticator(): Authenticator {
  return {
    type: 'ethTx',
    createProposeCall(envelope: Envelope<Propose>, args: ProposeCallArgs): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_propose', [
        space,
        args.author,
        {
          address: args.executionStrategy.address,
          params: [args.executionStrategy.params]
        },
        args.strategiesParams
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_propose',
        calldata: compiled
      };
    },
    createVoteCall(envelope: Envelope<Vote>, args: VoteCallArgs): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_vote', [
        space,
        args.voter,
        args.proposalId,
        args.choice,
        args.votingStrategies.map(strategy => ({
          index: strategy.index,
          params: [strategy.params]
        }))
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_vote',
        calldata: compiled
      };
    },
    createUpdateProposalCall(
      envelope: Envelope<UpdateProposal>,
      args: UpdateProposalCallArgs
    ): Call {
      const { space, authenticator } = envelope.data.message;

      const compiled = callData.compile('authenticate_update_proposal', [
        space,
        args.author,
        args.proposalId,
        {
          address: args.executionStrategy.address,
          params: [args.executionStrategy.params]
        }
      ]);

      return {
        contractAddress: authenticator,
        entrypoint: 'authenticate_update_proposal',
        calldata: compiled
      };
    }
  };
}
