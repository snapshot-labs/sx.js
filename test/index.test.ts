import { StarkNetTx, EthereumSig } from '../src/clients';
import { Account, defaultProvider, ec } from 'starknet';
import { Wallet } from '@ethersproject/wallet';

// Factory: 0x3e5165026e6586cd2d5cdf1fdced8af866f900e430bf1dc7b839c84604c506e
// Spaces
// [vanilla auth, vanilla voting strategy]
// 0x03ddbcdfec756eb9e9e0fbce7a90721570a09fc97d31a15b8e40769b93ecb957
// [ethSig auth, vanilla voting strategy]
// 0x03a48d8a7e80b7b295ef15efc4e1604e78b1f4460266cd460b6eb8826846bba0
// [ethSig auth, single slot proof (WETH goerli)]
// 0x069555971fbf76b3d0471297818ed93986fdd7afe3816d53ea8d8e72034260d8

describe('StarkNetTx', () => {
  const ethUrl = process.env.GOERLI_NODE_URL as string;

  const wallet = Wallet.createRandom();
  const walletAddress = wallet.address;
  const address = '0x00c26a3cdcc570da83f3dd6afd0db9d038ee096e2c56707d6348db3b06223427';
  const privKey = '0x78a61bf2d838b1094705168bbd5b462e665a5ce094d8b1c2e5438c66eeb9f59';
  const starkKeyPair = ec.getKeyPair(privKey);
  const account = new Account(defaultProvider, address, starkKeyPair);

  describe('vanilla authenticator', () => {
    const client = new StarkNetTx({ ethUrl });
    const space = '0x03ddbcdfec756eb9e9e0fbce7a90721570a09fc97d31a15b8e40769b93ecb957';
    const authenticator = '0x6ad07205a4d725c5c2b10c4f5fbdfaaa351c742fce7a5a22b2b56fd8d5afd62';
    const strategy = '0x344a63d1f5cd0e5f707fede9886d5dd306e86eba91ea410b416f39e44c3865';

    it('StarkNetTx.propose()', async () => {
      const envelope = {
        address: walletAddress,
        sig: null,
        data: {
          message: {
            space,
            authenticator,
            strategies: [strategy],
            metadataURI: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca',
            executionParams: []
          }
        }
      };

      const receipt = await client.propose(account, envelope);
      console.log('Receipt', receipt);

      // await defaultProvider.waitForTransaction(receipt.transaction_hash);

      expect(receipt.transaction_hash).toBeDefined();
    }, 60e3);

    it('StarkNetTx.vote()', async () => {
      const envelope = {
        address: walletAddress,
        sig: null,
        data: {
          message: {
            space,
            authenticator,
            strategies: [strategy],
            proposal: 1,
            choice: 1
          }
        }
      };

      const receipt = await client.vote(account, envelope);
      console.log('Receipt', receipt);

      // await defaultProvider.waitForTransaction(receipt.transaction_hash);

      expect(receipt.transaction_hash).toBeDefined();
    }, 60e3);
  });

  describe('ethSig authenticator', () => {
    const client = new StarkNetTx({ ethUrl });
    const ethSigClient = new EthereumSig(walletAddress);
    const space = '0x03a48d8a7e80b7b295ef15efc4e1604e78b1f4460266cd460b6eb8826846bba0';
    const authenticator = '0x594a81b66c3aa2c64577916f727e1307b60c9d6afa80b6f5ca3e3049c40f643';
    const strategy = '0x344a63d1f5cd0e5f707fede9886d5dd306e86eba91ea410b416f39e44c3865';

    it('StarkNetTx.propose()', async () => {
      const envelope = await ethSigClient.propose(wallet, walletAddress, {
        space,
        authenticator,
        strategies: [strategy],
        metadataURI: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca',
        executionParams: []
      });

      const receipt = await client.propose(account, envelope);
      console.log('Receipt', receipt);

      // await defaultProvider.waitForTransaction(receipt.transaction_hash);

      expect(receipt.transaction_hash).toBeDefined();
    }, 60e3);

    it('StarkNetTx.vote()', async () => {
      const envelope = await ethSigClient.vote(wallet, walletAddress, {
        space,
        authenticator,
        strategies: [strategy],
        proposal: 1,
        choice: 1
      });

      const receipt = await client.vote(account, envelope);
      console.log('Receipt', receipt);

      // await defaultProvider.waitForTransaction(receipt.transaction_hash);

      expect(receipt.transaction_hash).toBeDefined();
    }, 60e3);
  });

  describe('ethSig authenticator + single slot proof', () => {
    expect(process.env.GOERLI_NODE_URL).toBeDefined();
    expect(process.env.ETH_PK).toBeDefined();

    const wallet = new Wallet(process.env.ETH_PK as string);
    const walletAddress = wallet.address;

    const client = new StarkNetTx({ ethUrl });
    const ethSigClient = new EthereumSig(walletAddress);
    const space = '0x069555971fbf76b3d0471297818ed93986fdd7afe3816d53ea8d8e72034260d8';
    const authenticator = '0x594a81b66c3aa2c64577916f727e1307b60c9d6afa80b6f5ca3e3049c40f643';
    const strategy = '0x4bbd8081b1e9ef84ee2a767ef2cdcdea0dd8298b8e2858afa06bed1898533e6';

    it('StarkNetTx.getAccountProveCalls()', async () => {
      const envelope = await ethSigClient.propose(wallet, walletAddress, {
        space,
        authenticator,
        strategies: [strategy],
        metadataURI: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca',
        executionParams: []
      });

      const calls = await client.getProveAccountCalls(envelope, {
        strategyParams: ['0xB4FBF271143F4FBf7B91A5ded31805e42b2208d6', '0x3'],
        block: 7529615
      });

      expect(calls).toMatchSnapshot();
    });

    it('StarkNetTx.propose()', async () => {
      const envelope = await ethSigClient.propose(wallet, walletAddress, {
        space,
        authenticator,
        strategies: [strategy],
        metadataURI: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca',
        executionParams: []
      });

      const receipt = await client.propose(account, envelope);
      console.log('Receipt', receipt);

      // await defaultProvider.waitForTransaction(receipt.transaction_hash);

      expect(receipt.transaction_hash).toBeDefined();
    }, 60e3);

    it('StarkNetTx.vote()', async () => {
      const envelope = await ethSigClient.vote(wallet, walletAddress, {
        space,
        authenticator,
        strategies: [strategy],
        proposal: 5,
        choice: 1
      });

      const receipt = await client.vote(account, envelope);
      console.log('Receipt', receipt);

      // await defaultProvider.waitForTransaction(receipt.transaction_hash);

      expect(receipt.transaction_hash).toBeDefined();
    }, 60e3);
  });
});
