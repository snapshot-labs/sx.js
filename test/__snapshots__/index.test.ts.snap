// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarkNetTx ethSig authenticator + single slot proof StarkNetTx.getAccountProveCalls() 1`] = `
Array [
  Object {
    "calldata": Array [
      15,
      7529615,
      "0xb4fbf271143f4fbf",
      "0x7b91a5ded31805e4",
      "0x2b2208d6",
      7,
      "0x214",
      "0x214",
      "0x214",
      "0x214",
      "0x214",
      "0x174",
      "0x73",
      7,
      "0x43",
      "0x43",
      "0x43",
      "0x43",
      "0x43",
      "0x2f",
      "0xf",
      397,
      "0xf90211a07b26c06c",
      "0x38a71a1ecf4389fd",
      "0x2fb2d93e19080d0b",
      "0xfd4d76f9681df839",
      "0x10d84bb0a0bf1355",
      "0x91016ba0251a78bc",
      "0x37c3482c29777556",
      "0x5f75fd4d64d2795d",
      "0x5f6677a41da086c5",
      "0x9955eeb01b7595b7",
      "0xf171e7f1e472f02c",
      "0xd2593fa76a41cb4a",
      "0xe2e48d5fabb5a0ef",
      "0xc5d073788bdb8e40",
      "0x9765ca823fa5ab91",
      "0x40506f1083e9ac25",
      "0x0354ae6485c01ba0",
      "0x1ca687ceec241236",
      "0xab184911986bcdc6",
      "0x6dda1047e137e4ce",
      "0xefba9aa705f38505",
      "0xa0bb090d0a8e6926",
      "0x7bbcc4271382f02a",
      "0x2d753745a22b25de",
      "0x93469528bb4ca9e3",
      "0x85a0ba936482b6a7",
      "0x9ed3516c6d87db40",
      "0xf2e14208df189e01",
      "0x5a372580f7b9cf98",
      "0xe924a040ad81b909",
      "0x26608ad205336e4d",
      "0x45e9e4d1df5c0b54",
      "0xa1a921adda9288fb",
      "0x0dc2c1a0740c964f",
      "0xda7cbac68ebba1f6",
      "0xece96f7529d00fab",
      "0x60dac9ec8e2dc993",
      "0xebe14caaa062512a",
      "0xe7290558fb4674a5",
      "0xdf4ece677d8c1b56",
      "0x202ee1a311a15d42",
      "0x79ffce06eea00c11",
      "0x274b10a9877ce076",
      "0x7ecd43c3abbf8229",
      "0xecafac280283d4c3",
      "0x62c5385f5ca4a04d",
      "0x9ac440b85fc974be",
      "0xa50e0ad0a52964ca",
      "0x5d861d603f33af22",
      "0x27c9fee63f95eaa0",
      "0xf378b5c4a66ba8c4",
      "0xaea97b9da8fa11b3",
      "0xe84250eb1f6fe882",
      "0xd33b90dcbb01827c",
      "0xa08e5d63f046e28f",
      "0x8ef431ea1e45b315",
      "0xfb5f072271833e7d",
      "0x674965771fe8bbb5",
      "0xd7a094d1b9ff19ed",
      "0x1707002b6e0c6392",
      "0xf61168ac33463683",
      "0xae52e0067238b585",
      "0x5b51a0fd3a0be241",
      "0x84b536784c83dc6f",
      "0x3958985493c8b81a",
      "0x813577c408ed069c",
      "0x71d1c180",
      "0xf90211a062ea5dc3",
      "0x10f19b7b3734640e",
      "0x230893d7d5df1324",
      "0xfc7bab07e065690e",
      "0x9d98a4d4a08df7e8",
      "0xb58cf3182bdefca4",
      "0x4dcedf3a05128ded",
      "0x93856e4f062dee14",
      "0xdc3ee7d938a06ca5",
      "0x20c7effedc1dd739",
      "0x1bfd8bd78cf2d211",
      "0x67162543bd3ee4ad",
      "0x9593a11d881aa030",
      "0x4d8b7388a9a42b3b",
      "0x94ccefd18fd09b5e",
      "0xb3f0107cfa557e90",
      "0x00be936cc42557a0",
      "0xe15de0d7ab85785d",
      "0xab52df6396e9697a",
      "0x7b83bb0350ceeadf",
      "0x542fa42191dc066b",
      "0xa0380a1fa8425536",
      "0x7a7beb5d9e888cb7",
      "0x930292110fef8cd6",
      "0x455d2e75724c7ef5",
      "0x06a0be89961d3c1c",
      "0xb59ab32c1554a452",
      "0x0542e36a2abb00a0",
      "0x2f888f25ecbf58ee",
      "0x92e2a06b48cf6875",
      "0xb1220a4235e51e45",
      "0x2796ae95955c0f1c",
      "0x3d4fbfc9e0cecafb",
      "0xdb1e2da092f81df8",
      "0xca827ba3c9953efd",
      "0xd6904d9935f904db",
      "0x2c4b26cf4ba938f0",
      "0x4ad76d05a0b9c5b2",
      "0x4298b3adb8d1ae97",
      "0xdb8eec22223f474e",
      "0x3a765e068e1c369c",
      "0x15d4a8efdea0ac44",
      "0xf96c8d042953505e",
      "0xf0b7517c87e03ef7",
      "0xf447f3a5f7ab18bc",
      "0x0e9c5db1cd78a0e0",
      "0xf264377d3d55091d",
      "0x0e312c18da717dcb",
      "0x7ad91caaf15100f6",
      "0xd0b479ae4a1d24a0",
      "0x6d3120630c7f62b0",
      "0x10f1fe21dc8a2f34",
      "0x7475a59e728457b1",
      "0xb2f918c6ff76f34c",
      "0xa01ba103bd0c3681",
      "0x086e9d300cb2e76d",
      "0x306e556a525cb7cc",
      "0xe95bf9c072630bc7",
      "0x71a06e13965323b1",
      "0xbb4aea90ccbdae5a",
      "0xa29141a4aa53b37a",
      "0xabf4af844f098758",
      "0x8fe7a0567d59d922",
      "0x8f08fafc5a2e0e2f",
      "0x9d19ba7ac96cdb2a",
      "0x6f1cd7cc3bf9ca28",
      "0x55da3380",
      "0xf90211a0c28eb5e5",
      "0xac9b5d99c99ee3fb",
      "0x7c2121fbaa6f9236",
      "0x1218edfbcef0444a",
      "0xf54b3a59a0edd062",
      "0xe30a171902168951",
      "0x20416cffeeb9e4ad",
      "0x08a68801c61eb0c5",
      "0x9bd621ef3ca069dc",
      "0x559e14a3f9a88aa2",
      "0x72d1469fc2a87bfa",
      "0x121189cc08b75049",
      "0xab65043f1746a0bd",
      "0xfb4965b0e2bada0f",
      "0x4b129214cc971879",
      "0x96b985549b9d2890",
      "0xb17ff96eb4a8cfa0",
      "0x12ddc40066af5a30",
      "0x54eface27f043650",
      "0x4e40833ed10bba68",
      "0x3e8d6ece25bfa714",
      "0xa03bd89ba57e5439",
      "0x40df63308fecf31d",
      "0xe60c531b5868e820",
      "0x955e6ef9d2392aa4",
      "0x51a043fbb480d750",
      "0xe93d461e48f19393",
      "0x15d6134abd4b65ea",
      "0x3a9d52ebfdb704f8",
      "0x7bbaa0a71197f8ef",
      "0xe7c62af3c825435b",
      "0x0614029f110f4335",
      "0x3744d89035c70e46",
      "0x6e72d5a0942ea749",
      "0x5c8863a3d3e16716",
      "0xe22bae11608ceb47",
      "0xb1e2f84a72982b17",
      "0x4dfb47aba0defaf8",
      "0x809ad1f43e668576",
      "0x2f074b77bba5cd6f",
      "0x4f9411130f93d4e7",
      "0xc82e66d586a0a769",
      "0xb592220ee7dd57bc",
      "0xa7426c8ff989b76b",
      "0x9783f97ebe53ac07",
      "0xf9e777db7a22a0a5",
      "0x48da31a3839a706b",
      "0x508d52bea8b8c7eb",
      "0x78e0fa17fd897875",
      "0x453195f37bd053a0",
      "0xa743c955955df4c7",
      "0x1e43683530af7fbf",
      "0x27ed7b0da20baf0e",
      "0xbf2d06d569b8c7bf",
      "0xa07c3429a05963c3",
      "0x115beb8dfbc46a98",
      "0x0407209345237983",
      "0xe0ad193a39282716",
      "0x70a005f44c808b87",
      "0x318e010fab070965",
      "0x3e806b316aec5403",
      "0x2e76a50455e9dbc8",
      "0x86a3a02f08195b35",
      "0x672e9ed6b50930b0",
      "0x65d32c33c2a23e6b",
      "0x9316bb546a9a42a4",
      "0x1fabf080",
      "0xf90211a02e5ba7b0",
      "0x2466bca3a17c8efc",
      "0xc16467d24175c2b3",
      "0x51660a23e588b406",
      "0xc0802f89a04529f5",
      "0x0f23a222fc5e624c",
      "0xf38b1ce8e262562f",
      "0x0c51cecf3b037bc5",
      "0x38096d0a0ca097b1",
      "0xef134aeb9cbc4290",
      "0x3687c278b49f67d3",
      "0x19d875b1bf0e533e",
      "0x00984faf3dc3a0b9",
      "0x935cd611f6da7a2c",
      "0x64483dd9d18a922c",
      "0xd59b89b04902e384",
      "0xe9e27be7c8be9fa0",
      "0xaed38ea37f85bc53",
      "0x1bd40822e2087d97",
      "0x7802a4c3b5040162",
      "0xb063850c6fd319e2",
      "0xa091f95ab5943d38",
      "0x319d05d1d62ed512",
      "0xb2b829be9b967fbe",
      "0x7ae0bb1929ca68c8",
      "0xfea09edad4f10741",
      "0xb96065f7221854f7",
      "0x434249193e767f06",
      "0x5def83fc684a03d3",
      "0xd0daa077abc41fcf",
      "0xea5f03c7b3821478",
      "0x0a7b71235d9598aa",
      "0x6c5e54975e77cba7",
      "0x0fb821a0a27ae964",
      "0x66af1bd938d17b76",
      "0xf1fb494b68022f07",
      "0x6e61777894d13bcc",
      "0x0f9537eba0b6642c",
      "0x01ae021b5a9c1982",
      "0x067ce70d009f292e",
      "0x374d44866b7c348a",
      "0xedf79c467da01437",
      "0xccd5eb1df6d7f287",
      "0x019f14d13097fa9f",
      "0xb7f06dc871dc1dab",
      "0xc643746509afa0c4",
      "0xccd427a215fab496",
      "0x5f30bb3fff07e53c",
      "0x79b556149fd30d83",
      "0xd39a7d71f1a32ca0",
      "0x9c3f69de903167a8",
      "0x042369e420d5cefc",
      "0x03139ee0d7646c21",
      "0xb48dcb61ef8c1a5a",
      "0xa0189732d2b0a134",
      "0x2f373fedcdd841ca",
      "0x2abd926b19b47dfc",
      "0xfb25151729ec66d8",
      "0xdba069c736fb7ba8",
      "0x04ac94abc12b196b",
      "0x91c231326f261ba7",
      "0xdb803039d221bf1a",
      "0x9158a0835efef912",
      "0x8c92f54182cc1c2e",
      "0x6e2de760d1a276cf",
      "0x7dc3191ea86f2fd2",
      "0x4d5eac80",
      "0xf90211a09ebd199e",
      "0x9610b8f150e6a1ba",
      "0xe35878e8dded69a4",
      "0x41e4626befd927d4",
      "0x331b41a5a0e5f906",
      "0xbb34d649e44689bf",
      "0xf38e82ba2c87cb9c",
      "0xc81965afe8805725",
      "0x1adcd0a52fa00c57",
      "0x9884526e798e5120",
      "0xf3d66329b3cc8300",
      "0xa74eaffdbc9d0abc",
      "0x43c559712ef8a0cf",
      "0x99c0e31a0696a111",
      "0x25d7dbb8f18c569c",
      "0xc25cea2364748246",
      "0x0afcf9ed3311cca0",
      "0xd2798640e4354aee",
      "0x7ec677fe9a23318a",
      "0x67e8f60c1a57e867",
      "0x25f4bb122db412c1",
      "0xa019e79945ce5d3f",
      "0xc5bbd5e95497d75d",
      "0xc22f59801f1e7cb8",
      "0x299bf81a68cd7581",
      "0xd3a052497ea2f0a1",
      "0x0a26b88ab09d69fa",
      "0xb651dd1844ec1310",
      "0x6e8511583fcbe0b0",
      "0x8f40a04db561a777",
      "0xdd3da3f49d2a6a95",
      "0x48e3a033b182e131",
      "0xc52aaea37e4f8445",
      "0x32706da0e387e969",
      "0x8d9fa5b518161d79",
      "0x3ca477a47118f4f1",
      "0xc0ff25bc4fb388e9",
      "0xf1e36557a0f502e9",
      "0x811218090a5bfb2e",
      "0x0aeb94d7e9748030",
      "0x77c544d990da48d8",
      "0x06bf96e770a04be0",
      "0xd6907420a7823823",
      "0x1cbad51b962b4683",
      "0x06b1f5bc03dc682e",
      "0x4dae29acbd74a062",
      "0x74d240c6eab53080",
      "0x97a26071e820cf54",
      "0xea92cf780461e7ef",
      "0x6d93a86d830df1a0",
      "0x45e86c5b5ea86598",
      "0x56e408bb11f56e9e",
      "0xfbbdc266ec21cd6c",
      "0xd78870447eecb5a0",
      "0xa09724ffd336cf07",
      "0xdf9a9a3ac9feb7b4",
      "0x1d1b1379438a0664",
      "0xceb5afef58173f9e",
      "0x32a063f811b272d8",
      "0x46d6119ba3bf0051",
      "0xdebd7db00d1e928a",
      "0x1684fade501229e4",
      "0xd4a0a019f515756a",
      "0x70deaf72f912c9d9",
      "0xf5a17b3c67938309",
      "0xddb50637c496450e",
      "0xee29b980",
      "0xf90171a02fcb5ab3",
      "0xfdea4aecd8121a49",
      "0xfe16f13dab9cc242",
      "0x04779d29a292dbc0",
      "0xb2c75228a0d4448a",
      "0x82f6c9128162ae15",
      "0x08d1d373c4fde725",
      "0x2859f247d2bccc0d",
      "0xdaff99fe8da0322b",
      "0xe735938a0d7e5c6f",
      "0x70246dc6f5a9231a",
      "0xc252461e8788c97f",
      "0x80b311fc59a080a0",
      "0x53ac8d8823a1abdd",
      "0xb1bc30d2330b11ed",
      "0x4e82d7f597cd9215",
      "0xf930a7ead762b04d",
      "0x80a0a404bb21d51e",
      "0xfe336ad04e416cc7",
      "0x70c3d0f658ce20bc",
      "0xbdde0d72413903eb",
      "0x56bf8080a0034e99",
      "0xbb970d38fd7dd1aa",
      "0xca51bb7ed3bddd2a",
      "0xdd0b1b8eacc86990",
      "0xa80e0071b6a08586",
      "0xada2ef82ba5393d9",
      "0xbc874f4975296380",
      "0x59b62af37f5f2153",
      "0xeaa9980aaf6580a0",
      "0x6f335f8cf3c1bc9c",
      "0x3fae16fefa39aa8b",
      "0xaeb1bc537a81a674",
      "0x2e906f4bbb095889",
      "0xa0bbd2b83030d4d4",
      "0xe6a8f8bf17b44252",
      "0x12abde375f20e1f5",
      "0x4c7b36bd167f928b",
      "0x81a0f65fad8c91c1",
      "0x1e62922625a9e06c",
      "0xe6afc7bb6c9110c4",
      "0x008fe46d88dcc1cf",
      "0x5bdfa0bc2641ead2",
      "0x2cdff68ac20533e5",
      "0xbbd3b412b1d1a48c",
      "0x2be3a09627e10497",
      "0x3b638f80",
      "0xf8719e20294dda2d",
      "0x5979f269acd2630d",
      "0xff3880bca6301564",
      "0xc80e5af071862b4e",
      "0x26b850f84e018a45",
      "0x457742f07e48b96e",
      "0x2ca0e8bf5b62cbd9",
      "0x8b9280f253e3f54c",
      "0x440406363016ccee",
      "0x62f76e43ceabba6d",
      "0x1930a08b3eb75e29",
      "0xf2a9514d1a29784d",
      "0x34f29fd79929d886",
      "0x2345b0c70e467d57",
      "0x6b805d",
    ],
    "contractAddress": "0x363108ac1521a47b4f7d82f8ba868199bc1535216bbedfc1b071ae93cc406fd",
    "entrypoint": "prove_account",
  },
]
`;
